File: app/src/androidTest/java/com/github/sdpcoachme/firebase/auth/RealGoogleAuthenticatorTest.java
package com.github.sdpcoachme.firebase.auth;

import static android.app.Activity.RESULT_OK;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.junit.Assert.assertThrows;

import com.firebase.ui.auth.data.model.FirebaseAuthUIAuthenticationResult;

import org.junit.Test;

import java.util.concurrent.atomic.AtomicReference;

public class RealGoogleAuthenticatorTest {
    @Test
    public void onSignInWithFailedResultCodeExecutesFailCallBack() {
        RealGoogleAuthenticator realGoogleAuthenticator = new RealGoogleAuthenticator();
        AtomicReference<String> result = new AtomicReference<>("");

        realGoogleAuthenticator.onSignInResult(
                new FirebaseAuthUIAuthenticationResult(0, null)
                , (name) -> result.set("success")
                , () -> result.set("failure"));

        assertThat(result.get(), is("failure"));
    }

    @Test
    public void onSignInResultThrowsIfUserIsNull() {
        RealGoogleAuthenticator realGoogleAuthenticator = new RealGoogleAuthenticator();

        Exception error = assertThrows(IllegalStateException.class,
                () -> realGoogleAuthenticator.onSignInResult(
                        new FirebaseAuthUIAuthenticationResult(RESULT_OK, null)
                        , (name) -> {}
                        , () -> {}));
        assertThat(error.getMessage(), is("User is null"));
    }


}

File: app/src/androidTest/java/com/github/sdpcoachme/firebase/auth/RealGoogleAuthenticatorTest.kt
package com.github.sdpcoachme.firebase.auth

import android.app.Activity
import com.firebase.ui.auth.data.model.FirebaseAuthUIAuthenticationResult
import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert
import org.junit.Assert
import org.junit.Test
import java.util.concurrent.atomic.AtomicReference

class RealGoogleAuthenticatorTest {
    @Test
    fun onSignInWithFailedResultCodeExecutesFailCallBack() {
        val realGoogleAuthenticator = RealGoogleAuthenticator()
        val result = AtomicReference("")
        realGoogleAuthenticator.onSignInResult(
            FirebaseAuthUIAuthenticationResult(0, null),
            { result.set("success") },
            { result.set("failure") }
        )
        MatcherAssert.assertThat(result.get(), CoreMatchers.`is`("failure"))
    }

    @Test
    fun onSignInResultThrowsIfUserIsNull() {
        val realGoogleAuthenticator = RealGoogleAuthenticator()
        val error: Exception = Assert.assertThrows(
            IllegalStateException::class.java
        ) {
            realGoogleAuthenticator.onSignInResult(
                FirebaseAuthUIAuthenticationResult(Activity.RESULT_OK, null),
                {},
                {}
            )
        }
        MatcherAssert.assertThat(error.message, CoreMatchers.`is`("User is null"))
    }
}